Added `Properties` view

Corepex opened the pull request:
## Changes in this pull request
Resolves #182

Corepex committed:
first preperations for properties tab

Corepex committed:
Merge branch '1.x' into 182-properties

# Conflicts:
#	assets/js/src/core/modules/asset/editor/image/index.tsx
#	assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/preview/preview-container.styles.ts
#	assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/preview/preview-container.tsx
#	assets/js/src/core/modules/icon-library/index.ts

Corepex committed:
adjust api codegen settings according to new folder structure

Corepex committed:
added `element-toolbar`, splitted properties table into two, added api slice

Corepex committed:
fixed codegen command, fixed paths

Corepex committed:
added copy icon to value column, fixed flex style on table columns

Corepex committed:
added copy icon, reworked properties-container file

Corepex committed:
Merge branch '1.x' into 182-properties

# Conflicts:
#	translations/studio.en.yaml

Corepex committed:
added group icon to table, reworked columns due to new grid functions

Corepex committed:
added real world data to predefined properties select

Corepex committed:
added specific cell files for properties table

Corepex committed:
improved properties tab

Corepex committed:
added translations

Corepex committed:
build files

Corepex committed:
added several icons for property types

Corepex committed:
added icons to table, renamed `document-cell` to `element-cell`

Corepex committed:
build files

Corepex committed:
removed padding from text cell

Corepex committed:
build files

Corepex committed:
Merge branch '1.x' into 182-properties

# Conflicts:
#	assets/build/api/docs.jsonopenapi.json
#	assets/dist/core-dll/core-manifest.json
#	assets/dist/sdk/manifest.json
#	assets/js/src/core/modules/asset/asset-api-slice.gen.ts
#	public/core-dll/core-dll.048d5f5d.js.LICENSE.txt
#	public/core-dll/core-dll.30a5f688.js.LICENSE.txt
#	public/core-dll/core-dll.8f77c219.js.LICENSE.txt
#	public/core-dll/entrypoints.json
#	public/core-dll/manifest.json

Corepex committed:
updated api build

Corepex committed:
build files

vin0401 commented on assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/properties/components/table/cells/select-cell/select-cell.tsx:
```diff
@@ -0,0 +1,77 @@
+        onChange={ saveValue }
+        onKeyDown={ onKeyDown }
+        open={ open }
+        options={ [
+          { value: 'jack', label: 'Jack' },
+          { value: 'lucy', label: 'Lucy' },
+          { value: 'Yiminghe', label: 'yiminghe' }
+        ] }
```
Is it right that this is still hardcoded?
What are we missing here?

Corepex replied:
I was thinking about the translation of the stings. 
We only get a comma-separated list from the BE, something like "foo,bar,bazz." 
Wasn't sure how I should handle that - I implemented a basic display of the predefined options for now. (see https://github.com/pimcore/studio-ui-bundle/pull/324/commits/24cc73433484e68ddd6653de91fcacacabf06e2d)

@markus-moser, does Pimcore have an option to translate the options of a select in the predefined props?

markus-moser replied:
@Corepex In the classic UI it's not translatable. So let's leave it as not translateable for now.

Corepex replied:
Thanks for checking @markus-moser. I guess than the current implementation should do it :)

vin0401 commented on assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/properties/components/table/cells/select-cell/select-cell.tsx:
```diff
@@ -0,0 +1,77 @@
+      saveValue(e.target.value)
+    }
+
+    function onKeyDown (e: React.KeyboardEvent<HTMLInputElement>): void {
```
Would be nice when we also could add the enter key here to save the value.
Like we do it for the text field.

Corepex replied:
Good point. Implemented with https://github.com/pimcore/studio-ui-bundle/pull/324/commits/0f835c21c2a1154f35528d784d7020b665836a1f

vin0401 commented on assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/properties/components/table/cells/boolean-cell/boolean-cell.tsx:
```diff
@@ -0,0 +1,71 @@
+    return (
+      <Checkbox
+        checked={ props.getValue() }
+        onChange={ saveValue }
```
I would only send the update for the data when hitting enter.
Currently the value would directly send via update event even so when I try to abort with escape.

So maybe we use the original cell value just as an initial value for the component state. 
Hitting enter then sends the state value per update event.

Corepex replied:
Not sure about the usability here. My idea was to send the changes each time you check the checkbox (because obv. it is a change). Don`t know if the "send on enter" conflicts with the keyboard navigation.

vin0401 commented on assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/properties/components/table/cells/select-cell/select-cell.tsx:
```diff
@@ -0,0 +1,77 @@
+/**
```
Maybe we can move it to the grid folder?
For now this looks quite reusable for different types of grids. :)

Corepex replied:
done with https://github.com/pimcore/studio-ui-bundle/pull/324/commits/458d218286356d64030e6c25af5c0d3fe91beeaf

vin0401 commented on assets/js/src/core/modules/asset/editor/image/tab-manager/tabs/properties/components/table/cells/boolean-cell/boolean-cell.tsx:
```diff
@@ -0,0 +1,71 @@
+/**
```
Maybe we can move this file to the grid folder. For now this looks quite reusable. :)

Corepex replied:
done with https://github.com/pimcore/studio-ui-bundle/pull/324/commits/458d218286356d64030e6c25af5c0d3fe91beeaf

vin0401 reviewed (changes requested):
Looks good already. I just have some small improvements I would suggest, here. :)

markus-moser commented:
@Corepex Things I noticed during testing:

- [x]  Adding custom properties is not yet implemented. The properties can be saved later together with the asset, so the add button should just add the new row to the table/form:

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/941348f7-29ec-41fe-9944-67912d2bf2e8)

- [x]  Adding predefined properties not implemented yet (same like custom properties):
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/d50ebba9-cf33-480e-b7bf-60b9400812f2)

- [x]  As the "All Properties" tab is a ready-only view I would remove the "predefined properties" select and "Add a custom property" button there and show it only in the "edit own properties" tab.
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/db461ebe-6da5-47bb-8229-02efc4539847)

- [x] The properties tab is currently only visible for image assets. It should be visible for folders, text assets etc. too

- [x] When I close and re-open an asset the properties information get's not reloaded from the server. We need to refresh the data on each (re-)load.

- [x] When I edit the value of for example a text field the value get's not updated in the table (not sure if related to this PR or a "problem" of the inline edit functionality itself)

- [x] The open button currently is not implemented yet:

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/379a7719-ea14-41ef-a113-c0a4aab8a2c5)

- [x] The delete button is missing in the "Actions" column:
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/9617606b-dc00-49e2-b031-1fa969e2589a)

Should it really be an "Actions" column if we display the "open" button directly in the "value" column and therefore only the "Delete" action is present. The column should be called "Delete", right?

- [x] Inline edit of (inheritable) checkboxes currently not implemented:

- [x] The "All Properties" tab is a readonly view. Therefore the inline edit functionallity needs to be disabled and the "Actions" column should be hidden.
- [x] (Nice2Have) When the properties list is empty the "All properties" view looks a little bit strange, do we have an idea how to improve it?

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/1a9ef2f0-ff2e-4d19-982e-29110411e188)

- [x] We need to find a way for our table grid to handle long content:
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/75c8e815-80c3-462d-8111-34357027965a)

It does not make sense to make the tables that narrow when we have a lot of content. Don't know what the best approaches are, but we should use the width of the panel if the content requests it.

- [x] Should the checkbox value column really be centered? For me this looks a bit strange in combination with other types:
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/bed0fe78-4d0f-482c-b59a-57b323e7e8e0)

- [x] The "Description" column must not be editable

- [x] The "Description" column should display the word "Custom" for custom properties

- [x] "Select" property options are hardcoded:
![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/50aeb19f-4181-4f95-aa8c-bce10d8bae89)

- [x] "Select type for properties has no icon:

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/1c63ac5e-d463-40b8-a22a-abfeee47760c)

Please let me know if I should split this up into multiple issues...

Corepex committed:
added real options to select in select-cell

Corepex committed:
added `enter` to `onKeyDown` triggers

Corepex committed:
moved `boolean` and `select` cell to grid component folder

Corepex commented:
@markus-moser according your changes:

**Adding custom properties is not yet implemented. The properties can be saved later together with the asset, so the add button should just add the new row to the table/form.**

TBI


**Adding predefined properties not implemented yet (same like custom properties).**

TBI


**As the "All Properties" tab is a ready-only view I would remove the "predefined properties" select and "Add a custom property" button there and show it only in the "edit own properties" tab.**

removed the mentioned buttons on "all" tab. (see https://github.com/pimcore/studio-ui-bundle/pull/324/commits/5e071cbb62a44035169f4e8180a63dc6cd92cbb9)


**The properties tab is currently only visible for image assets. It should be visible for folders, text assets etc. too**

TBI



**When I close and re-open an asset the properties information get's not reloaded from the server. We need to refresh the data on each (re-)load.**

TBI


**When I edit the value of for example a text field the value get's not updated in the table (not sure if related to this PR or a "problem" of the inline edit functionality itself)**

TBI


**The open button currently is not implemented yet:**

True, that's because we can potentially have links to `documents` or `dataobjects` and this will break the application no - should we only implement it for assets (for now)?


**The delete button is missing in the "Actions" column**

changed icon in https://github.com/pimcore/studio-ui-bundle/pull/324/commits/16984ea8f7a78a25f97494a62011b3dc01da9708 according to #326 


**Inline edit of (inheritable) checkboxes currently not implemented**

TBI


**The "All Properties" tab is a readonly view. Therefore the inline edit functionallity needs to be disabled and the "Actions" column should be hidden.**

TBI


**(Nice2Have) When the properties list is empty the "All properties" view looks a little bit strange, do we have an idea how to improve it?**

The table will only be shown if there at least one entry. (see https://github.com/pimcore/studio-ui-bundle/pull/324/commits/a3247e83897b83d1a03d6a53fab235ff811aaad4)


**We need to find a way for our table grid to handle long content:**

I already talked to @vin0401 about the grid scaling issue. There is also no option for now to make the grid full-width.
https://github.com/pimcore/studio-ui-bundle/pull/324/commits/e7a90477e4b7ab34dee512c7b5787111a4484909


**Should the checkbox value column really be centered? For me this looks a bit strange in combination with other types:**

IMHO It also looks weird when its only on the left ... don't know what the best solution is ...


**The "Description" column must not be editable**

see https://github.com/pimcore/studio-ui-bundle/pull/324/commits/5f4a4b028b1962736b31900016148e0f68cc135a


**The "Description" column should display the word "Custom" for custom properties**

Added the `name` column. (https://github.com/pimcore/studio-ui-bundle/pull/324/commits/d8c9417332ef60247c29de7cbd55dd01624bff7f, https://github.com/pimcore/studio-ui-bundle/pull/324/commits/36c8920700a05d47ba46c4a490e1931668baed71)


**"Select" property options are hardcoded:**

please see: https://github.com/pimcore/studio-ui-bundle/pull/324#discussion_r1621948480


**"Select type for properties has no icon:**

That is true; this icon has to be redesigned by @idaiv. - added with https://github.com/pimcore/studio-ui-bundle/pull/324/commits/8ce9702228d903ce453c54796e7474f2bfd4a524 and https://github.com/pimcore/studio-ui-bundle/pull/324/commits/4425112e0eb939135427a34c81d45f1d7192604d

https://github.com/pimcore/studio-ui-bundle/pull/326

Corepex committed:
add predefined properties only visible in "own" mode

Corepex committed:
display tables only if they have at least one entry

Corepex committed:
removed the ability to edit the description column

vin0401 commented:
@Corepex - In regards to this point:

> We need to find a way for our table grid to handle long content:
>
>I already talked to @vin0401 about the grid scaling issue. There is also no option for now to make the grid full-width.

We can't make it full width but at least we can add some additional handling for long content.
And we already have that for the asset listing. 

![image](https://github.com/pimcore/studio-ui-bundle/assets/26813978/28097d0b-736b-4487-bc0b-273b2c45d1e2)

I guess there is a new wrapper container that breaks the css in that case. But a few adaptions done to this container should already fix that. Reach out to me when you need help. :)

idaiv commented:
to @markus-moser points:
- Should it really be an "Actions" column if we display the "open" button directly in the "value" column and therefore only the "Delete" action is present. The column should be called "Delete", right? ---> yes, when there is a single action like Delete, then the column label is based on it. If there are multiple actions, then the label is "Actions"
- (Nice2Have) When the properties list is empty the "All properties" view looks a little bit strange, do we have an idea how to improve it? ---> We can add a message for all empty tables in general "No context/data yet." to inform the user. And still keep the column headings for a reference what content is supposed to be in such a table. 
- "Select type for properties has no icon ---> yes, we have discussed with Christian and couldnt find an example on the current UI of such. I should create a new one without a reference then? 

And Markus, please let me know if I missed anything from the list or you need certain examples.

markus-moser commented:
@idaiv Thanks for your input 👍 

The select icon looks like this in the current UI:

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/f837fecb-19a4-41b3-a576-512f2191831b)

Maybe you have an idea for the checkbox topic?

![image](https://github.com/pimcore/studio-ui-bundle/assets/4639428/7fca18b8-7336-4621-8a97-e9cb0a5ea852)

Corepex committed:
added `name` column to properties table

Corepex committed:
improved styling of property table columns

Corepex committed:
fixed delete icon

Corepex committed:
added translations file

Corepex committed:
added icon for `select` property

Corepex committed:
added icon for `select` property v2

Corepex committed:
changed icon accoding trash icon PR

Corepex committed:
corrected `trash` icon

Corepex committed:
simplified `useAsset` hoook

Corepex committed:
improved path cell in `value` cell

Corepex committed:
made checkboxes clickable

Corepex committed:
fixed tag title

Corepex committed:
Merge branch '1.x' of github.com:pimcore/studio-ui-bundle into 182-properties

markus-moser commented:
@Corepex @vin0401 I added follow-up tickets for the remaining properties topics - let's continue our work based on those issues:

#329 
#330 
#331 
#332 
#333 
#334
